adder/adder.go

```diff
@@ -19,20 +19,33 @@ import (
 	files "github.com/ipfs/boxo/files"
 	merkledag "github.com/ipfs/boxo/ipld/merkledag"
 	cid "github.com/ipfs/go-cid"
-	cbor "github.com/ipfs/go-ipld-cbor"
 	ipld "github.com/ipfs/go-ipld-format"
+	ipldlegacy "github.com/ipfs/go-ipld-legacy"
 	logging "github.com/ipfs/go-log/v2"
+	dagpb "github.com/ipld/go-codec-dagpb"
+	"github.com/ipld/go-ipld-prime/codec/dagcbor"
+	"github.com/ipld/go-ipld-prime/codec/raw"
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+	"github.com/ipld/go-ipld-prime/multicodec"
+	"github.com/ipld/go-ipld-prime/node/basicnode"
 	multihash "github.com/multiformats/go-multihash"
 )
 
 var logger = logging.Logger("adder")
 
-// go-merkledag does this, but it may be moved.
-// We include for explicitness.
+var ipldDecoder *ipldlegacy.Decoder
+
+// create an ipld registry specific to this package
 func init() {
-	ipld.Register(cid.DagProtobuf, merkledag.DecodeProtobufBlock)
-	ipld.Register(cid.Raw, merkledag.DecodeRawBlock)
-	ipld.Register(cid.DagCBOR, cbor.DecodeBlock)
+	mcReg := multicodec.Registry{}
+	mcReg.RegisterDecoder(cid.DagProtobuf, dagpb.Decode)
+	mcReg.RegisterDecoder(cid.Raw, raw.Decode)
+	mcReg.RegisterDecoder(cid.DagCBOR, dagcbor.Decode)
+	ls := cidlink.LinkSystemUsingMulticodecRegistry(mcReg)
+
+	ipldDecoder = ipldlegacy.NewDecoderWithLS(ls)
+	ipldDecoder.RegisterCodec(cid.DagProtobuf, dagpb.Type.PBNode, merkledag.ProtoNodeConverter)
+	ipldDecoder.RegisterCodec(cid.Raw, basicnode.Prototype.Bytes, merkledag.RawNodeConverter)
```

Comment:
Needs a decision on the type of registry to use here:
- go-ipld-prime codecs with wrappers for dag-pb + raw ( e.g. the initial proposal here )
- go-ipld-prime globals (e.g. including other codecs) with dag-pb + raw wrappers
- go-ipld-format only registry

cc @hsanjuan

Reply:
What do you recommend?

Reply:
I'm not sure how the performance of the codecs compares (e.g. the two dag-cbor codecs). I feel like I've heard both are better from different parties.

I'd stay away from global variables though, since less magic makes it more obvious what you do/don't support.

---

Whether you go with the go-ipld-format codecs and then add go-ipld-legacy wrappers for go-ipld-prime codecs, or use go-ipld-prime codecs with the couple necessary wrappers for dag-pb/raw doesn't make a big difference. I've heard people say the old codecs are faster, and I've heard the inverse too. Haven't had the motivation to benchmark and see though 🤷.

Reply:
I guess we can take what you did here then :shrug: